tools/PI/DevHome.PI/Helpers/SnapHelper.cs

```diff
@@ -0,0 +1,148 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System;
+using System.ComponentModel;
+using System.Diagnostics;
+using CommunityToolkit.Mvvm.ComponentModel;
+using DevHome.Common.Extensions;
+using DevHome.PI.Models;
+using DevHome.PI.ViewModels;
+using Microsoft.UI.Xaml;
+using Windows.Graphics;
+using Windows.Win32;
+using Windows.Win32.Foundation;
+using Windows.Win32.UI.Accessibility;
+using WinUIEx;
+
+namespace DevHome.PI.Helpers;
+
+public class SnapHelper
+{
+    // TODO The SnapOffsetHorizontal and UnsnapGap values don't allow for different DPIs.
+    private const int UnsnapGap = 9;
+
+    // It seems the way rounded corners are implemented means that the window is really 8px
+    // bigger than it seems, so we'll subtract this when we do sidecar snapping.
+    private const int SnapOffsetHorizontal = 8;
+
+    private readonly WINEVENTPROC _winPositionEventDelegate;
+    private readonly WINEVENTPROC _winFocusEventDelegate;
+
+    private HWINEVENTHOOK _positionEventHook;
+    private HWINEVENTHOOK _focusEventHook;
+
+    public SnapHelper()
+    {
+        _winPositionEventDelegate = new(WinPositionEventProc);
+        _winFocusEventDelegate = new(WinFocusEventProc);
+    }
+
+    public void Snap()
+    {
+        Debug.Assert(_positionEventHook == HWINEVENTHOOK.Null, "Hook should be cleared");
+        Debug.Assert(_focusEventHook == HWINEVENTHOOK.Null, "Hook should be cleared");
+
+        _positionEventHook = WindowHelper.WatchWindowPositionEvents(_winPositionEventDelegate, (uint)TargetAppData.Instance.ProcessId);
+        _focusEventHook = WindowHelper.WatchWindowFocusEvents(_winFocusEventDelegate, (uint)TargetAppData.Instance.ProcessId);
+
+        SnapToWindow();
+    }
+
+    public void Unsnap()
+    {
+        var barWindow = Application.Current.GetService<PrimaryWindow>().DBarWindow;
+        Debug.Assert(barWindow != null, "BarWindow should not be null.");
+
+        // Set a gap from the associated app window to provide positive feedback.
+        PInvoke.GetWindowRect(barWindow.CurrentHwnd, out var rect);
+        barWindow.UpdateBarWindowPosition(new PointInt32(rect.left + UnsnapGap, rect.top));
+
+        if (_positionEventHook != HWINEVENTHOOK.Null)
+        {
+            PInvoke.UnhookWinEvent(_positionEventHook);
+            _positionEventHook = HWINEVENTHOOK.Null;
+        }
+
+        if (_focusEventHook != HWINEVENTHOOK.Null)
+        {
+            PInvoke.UnhookWinEvent(_focusEventHook);
+            _focusEventHook = HWINEVENTHOOK.Null;
+        }
+    }
+
+    private void WinPositionEventProc(HWINEVENTHOOK hWinEventHook, uint eventType, HWND hwnd, int idObject, int idChild, uint idEventThread, uint dwmsEventTime)
+    {
+        // Filter out events for non-main windows.
+        if (idObject != 0 || idChild != 0)
+        {
+            return;
+        }
+
+        if (hwnd == TargetAppData.Instance.HWnd)
```

Comment:
Consider using early returns instead of deep nesting.

if hwd does not equal the target hwnd, return.